### ISSUE TITLE
ZIO Test: `@@ only` TestAspect

### ISSUE THREAD

**Issue by reibitto** [zio/zio#3226]
This idea came up once before here:
https://github.com/zio/zio/pull/1838#issuecomment-535629515

Basically for when you want to run individual tests/suites with a TestAspect.

I like this idea for cases where you don't want to deal with labels and such.

There might be a few corner cases to think about when it comes to precedence. For example, what happens when you have an `@@ ignore` on a suite, but an `@@ only` on a single test (and the reverse and so on).

**Comment by adamgfraser** [comment]
I think we have the tools to do this now. We can add a static test annotation with `@@ only` and modify the test executor to first scan the specs for ones annotated with `only` and if any exist, only run those. As you say there are some tricky cases where we basically give the test framework conflicting instructions but we should be able to figure something out there. I will work on this.

**Comment by nartamonov** [comment]
Release notes for RC19 mention `TestAspect.only` was implemented but actually it have been deleted from library in #3696. It' sad... Should we re-open the issue? It's very helpful combinator, at least for me.

**Comment by adamgfraser** [comment]
Yes, reopening to track it. Unfortunately this combinator is not resource safe with the current encoding of specs because we need to evaluate all the specs to determine whether there are any that are annotated with `only` so we can't close resources for a suite before evaluating the next suite. I'm working on an alternative encoding of the resources in specs that should allow us to bring this back but it is still work in process at this point.

**Comment by jypma** [comment]
Would it perhaps be easier (in the meantime) to have a simple environment variable control which spec(s) to run, e.g. just by name? Even if the skipped ones get their environment/dependencies constructed, it'd still save time and log output on that test run.

**Comment by adamgfraser** [comment]
@jypma You can already filter for a certain test with `testOnly zio.ZIOSpec -- -t "fiber interruption"`.

**Comment by adamgfraser** [comment]
Nice!

**Comment by kitlangton** [comment]
> Yes, reopening to track it. Unfortunately this combinator is not resource safe with the current encoding of specs because we need to evaluate all the specs to determine whether there are any that are annotated with `only` so we can't close resources for a suite before evaluating the next suite. I'm working on an alternative encoding of the resources in specs that should allow us to bring this back but it is still work in process at this point.

Just curious if you think this is feasible now? :D

**Comment by spavikevik** [comment]
Also interested to hear if there's any updates on this

**Comment by reibitto** [comment]
One other thing for this that we might have to be careful of is to have some way of preventing accidental commits of `TestAspect.only`. As far as I see it, it's mostly a local development thing. Ideally I think it should fail in CI. Or at least have a parameter/option for indicating that.

**Comment by 987Nabil** [comment]
I think this is more a foot gun then a feature. And there are ways to run single tests via sbt and the intellij plugin.
@reibitto I think your idea of a more configurable `-t` is actually the better one.

**Comment by reibitto** [comment]
Yeah it might be safer, but making `-t` more customizable seems like it might be rather complex and have its own downsides. Like you could create special syntax for picking out elements in a tree (for nested specs/tests) but I don't know... that feels like it could be overkill/complicated. And I wouldn't like it if it now required escaping special characters like `*` or whatever if you want to support suffix/prefix matching and so on for the other special characters that might be introduced (e.g. CSS selectors, xpath, etc). That feels like it would be annoying because now you can't copy/paste test names all the time and have it always work.

If the solution is to go with a `-t` like approach, then I almost think it should be an entirely new argument. Like `-x` or whatever. That way the selector syntax wouldn't be forced on you.

I don't use the built-in IntelliJ feature for various reasons. One is that I don't always see the test being recognized. Another is that I don't build from IntelliJ itself (or its sbt shell feature). It's never mixed well with my external terminal where sbt is running as an interactive process.

I don't feel too strongly either way if it's not under `-t`. The `only` feature is just more familiar to me because other test frameworks have it too, like munit.

**Comment by reibitto** [comment]
Or if it has to be shared with `-t` then maybe a modifier flag could work? Like `-t ... --selector` (ignore the poor naming, just an example 😅 )

**Comment by reibitto** [comment]
By the way, I just checked and yeah munit also has a CI check to prevent what I described. Which makes sense. I figured it would work that way.
